### DISCUSSION TITLE
sendmsg control headers, do they work in android ver 26, how do i find this out?

### DISCUSSION THREAD

**Discussion by timprepscius** [android/ndk#1929] [Q&A] [answered]
Greetings,

My question has two parts:
1.  do sendmsg msghdr msg_controllen msg_control work in android ver 26
3. how can I definitively determine what is and what is not supported?

Notes:
I'm trying to determine why a quic implementation is failing on android, while working ios/linux/mac/windows.
the sendmsg function, when msg_controllen > 0, return -1 with errno EINVAL.

**Comment by enh-google** [comment]
this is really a kernel question, and will most likely depend on what
kernel the device(s) in question have rather than the OS version, and
probably also on things like what their maximum msg_controllen is
configured to. (if you can read /proc/sys/net/core/optmem_max on the
device(s) in question, it might be worth a look at that.)

**Comment by timprepscius** [comment]
Is there a place I can look for this type of information?

I feel like there should be some document that I can look at instead of asking on random github discussion forums.

**Reply by enh-google** [reply]
> Is there a place I can look for this type of information?

depends on the device. i'd start by looking at the Android Common Kernel source (or just looking at "what's the kernel version(s) on the devices that are failing for me, and what's the lowest kernel version i see on a device that _does_ work").

but if you're only seeing it on devices from a particular OEM or SoC vendor, it's possible that the behavior difference is in that _specific_ kernel (and may not even be reflected upstream).

> I feel like there should be some document that I can look at instead of asking on random github discussion forums.

the linux kernel is quite a large API surface. when there's an actual bug that's caused widespread trouble, we'll typically add a CTS test. but where there's no test (and there's certainly no _direct_ test for sendmsg() control messages) there's pretty wide latitude. in this specific case, there are some parts of the system that use control messages.

> Also, are you suggesting that I post to the linux kernel mailing list - that doesn't seem right.

no, because they'll just say the same thing --- "what exact kernel(s)?", and then tell you to talk to the maintainers of those kernels.

> Is there a discussion group for android/linux kernel?

not that i'm aware of, no.

**Reply by timprepscius** [reply]
I know the kernel, it is "Linux version 3.10.84-perf-g51b663cc723-04742-g5bcfaca9e61"

So- just to verify, you are suggesting that I go and read the linux kernel source, in order to find out what kernel version supports sendmsg controllen?

**Comment by timprepscius** [comment]
Also, are you suggesting that I post to the linux kernel mailing list - that doesn't seem right.  Is there a discussion group for android/linux kernel?

**Comment by enh-google** [comment, marked as answer]
> I know the kernel, it is "Linux version 3.10.84-perf-g51b663cc723-04742-g5bcfaca9e61"
> 
> So- just to verify, you are suggesting that I go and read the linux kernel source, in order to find out what kernel version supports sendmsg controllen?

if it's not documented in the man page, reading the source is the only alternative i know of. (if you find anything interesting but don't want to go through the bureaucracy of trying to get the man page improved, let me know and i can try to upstream it for you.)

**Reply by timprepscius** [reply]
While I was quite incredulous, I looked up the source for linux kernel for the revision on this dev phone, read the networking stack code, and walah, the problem revealed itself.  I'm still not sure how to feel about this.
